webrtc/README.md

```diff
@@ -0,0 +1,423 @@
+  WebRTC connection from a remote node without previously receiving an SDP
+  packet from such host?
+
+### Browser to Browser
+
+Scenario: Browser _A_ wants to connect to Browser node _B_ with the help of
+server node _R_.
+
+- Replace STUN with libp2p's identify and AutoNAT
+  - https://github.com/libp2p/specs/tree/master/identify
+  - https://github.com/libp2p/specs/blob/master/autonat/README.md
+- Replace TURN with libp2p's Circuit Relay v2
+  - https://github.com/libp2p/specs/blob/master/relay/circuit-v2.md
+- Use DCUtR over Circuit Relay v2 to transmit SDP information
+  1. Transform ICE candidates in SDP to multiaddresses.
+  2. Transmit the set of multiaddresses to the remote via DCUtR.
+  3. Transform the set of multiaddresses back to the remotes SDP.
+  4. https://github.com/libp2p/specs/blob/master/relay/DCUtR.md
+
+#### Open Questions
+
+- Can a browser know upfront its UDP port which it is listening for incoming
+  connections on? Does the browser reuse the UDP port across many WebRTC
+  connections? If that is the case one could connect to any public node, with
+  the remote telling the local node what port it is perceived on.
+
+  No, a browser uses a new UDP port for each `RTCPeerConnection`.
+
+- Can _Browser_ control the lifecycle of its local TLS certificate, i.e. can
+  _Browser_ use the same TLS certificate for multiple WebRTC connections?
+
+  Yes. For the lifetime of the page, one can generate a certificate once and
+  reuse it across connections. See also
+  https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/RTCPeerConnection#using_certificates
+
+- Can two _Browsers_ exchange their SDP packets via a third server node using
+  Circuit Relay v2 and DCUtR? Instead of exchanging the original SDP packets,
+  could they exchange their multiaddr and construct the remote's SDP packet
+  based on it?
+
+- Instead of using trickle ICE, we could as well wait for the candidate
+  gathering. See
+  https://github.com/pion/webrtc/blob/c1467e4871c78ee3f463b50d858d13dc6f2874a4/examples/insertable-streams/main.go#L141-L142
+  as one example. In the browser, one can wait for the
+  [`icegatheringstatechange`
+  event](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/icegatheringstatechange_event).
+
+## Multiplexing
+
+The WebRTC browser APIs do not support half-closing of streams nor resets of the
+sending part of streams.
+[`RTCDataChannel.close()`](https://developer.mozilla.org/en-US/docs/Web/API/RTCDataChannel/close)
+flushes the remaining messages and closes the local write and read side. After
+calling `RTCDataChannel.close()` one can no longer read from the channel. This
+lack of functionality is problematic, given that libp2p protocols running on top
+of transport protocols, like WebRTC, expect to be able to half-close or reset a
+stream. See [Connection Establishment in
+libp2p](https://github.com/libp2p/specs/blob/master/connections/README.md#definitions).
```

Comment:
As an FYI, the linked document does not mention a "reset" capability.

Reply:
Can someone explain to me, in which situations we want to reset a stream?

From looking at the mplex implementation, a reset seems to be a non-graceful close of a stream, i.e. in Rust-land when we `Drop` a stream without calling `close` before. In a GC-language, this would be equivalent to not closing the stream but having it garbage-collected.

This is our only application of resets?

@mxinden In https://github.com/melekes/rust-libp2p/pull/8, you added support for handling reset frames but never send one. Was that intentional?
@marten-seemann In which circumstances does the go implementation reset the stream?

Does https://github.com/mxinden/specs/commit/65a894a0d2c57b37f639f4d68ae28d9ea03be481 perhaps mean that we don't need a reset at all?

Reply:
> > From looking at the mplex implementation, a reset seems to be a non-graceful close of a stream, i.e. in Rust-land when we `Drop` a stream without calling `close` before. In a GC-language, this would be equivalent to not closing the stream but having it garbage-collected.
> > This is our only application of resets? Is this the intended application of resets?
> 
> Yes, that is my mental model of a stream reset. In addition we might want to allow users to actively reset a stream, similar to your suggestion in [#412 (comment)](https://github.com/libp2p/specs/pull/412#discussion_r983113010) via an additional trait. Though here again, I suggest we only provide such a mechanism once there is the need for it.

Is this specified anywhere? The [search results](https://github.com/libp2p/specs/search?q=reset) for reset in this repository don't turn up anything in that regard. Is this just considered prior knowledge? :)

> > @mxinden In [melekes/rust-libp2p#8](https://github.com/melekes/rust-libp2p/pull/8), you added support for handling reset frames but never send one. Was that intentional?
> 
> No, that was not intentional. Sorry for not making this explicit. rust-libp2p should send a reset when dropping a stream.

Cool, I'll add that as a TODO for the implementation and will think about how we can cover this in our test suite.

Reply:
> > From looking at the mplex implementation, a reset seems to be a non-graceful close of a stream, i.e. in Rust-land when we `Drop` a stream without calling `close` before. In a GC-language, this would be equivalent to not closing the stream but having it garbage-collected.
> > This is our only application of resets? Is this the intended application of resets?
> 
> Yes, that is my mental model of a stream reset. In addition we might want to allow users to actively reset a stream, similar to your suggestion in [#412 (comment)](https://github.com/libp2p/specs/pull/412#discussion_r983113010) via an additional trait. Though here again, I suggest we only provide such a mechanism once there is the need for it.

If that is the case, we should fix the docs on the protobuf flag. That one currently only mentions the "writing side" of the stream but with this mental model, the "reading side" will obviously also be gone. After doing so more research, it seems to me that resets typically mean the connection was completed aborted. The [QUIC RFC](https://www.rfc-editor.org/rfc/rfc9000.html#frame-reset-stream) states that too unless I misunderstand something:

> An endpoint that receives a RESET_STREAM frame for a send-only stream MUST terminate the connection with error STREAM_STATE_ERROR.

"Terminate the connection" means it discard it completely yeah? Not just the sending half. In other words, reset is a terminal state.

Reply:
> After doing so more research, it seems to me that resets typically mean the connection was completed aborted. The [QUIC RFC](https://www.rfc-editor.org/rfc/rfc9000.html#frame-reset-stream) states that too unless I misunderstand something:
>
> > An endpoint that receives a RESET_STREAM frame for a send-only stream MUST terminate the connection with error STREAM_STATE_ERROR.

Since RESET_STREAM is only used by the _sender_, it's **invalid** to send a RESET_STREAM frame for a unidirectional _receive stream_. Unfortunately the QUIC frame format allows _encoding_ of such a RESET_STREAM frame, which is why this check is needed. Basically, if you receive such a RESET_STREAM frame, you know that your peer is severely misbehaving, and you immediately kill the connection.

Reply:
Gotcha, that clarifies the "send-only" bit. The overall conclusion remains though right? If I receive a `RESET` flag, I can consider the stream dead, i.e. there is no point in writing to it and expecting the other party to read that data.

Reply:
I opened an issue so we can discuss this outside of here: https://github.com/libp2p/specs/issues/466